src/main/java/slimeknights/tconstruct/plugin/jei/ToolBuildingCategory.java

```diff
@@ -0,0 +1,119 @@
+import mezz.jei.api.gui.ingredient.IRecipeSlotsView;
+import mezz.jei.api.helpers.IGuiHelper;
+import mezz.jei.api.recipe.IFocusGroup;
+import mezz.jei.api.recipe.RecipeIngredientRole;
+import mezz.jei.api.recipe.RecipeType;
+import mezz.jei.api.recipe.category.IRecipeCategory;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import slimeknights.tconstruct.TConstruct;
+import slimeknights.tconstruct.library.client.GuiUtil;
+import slimeknights.tconstruct.library.recipe.tinkerstation.building.ToolBuildingRecipe;
+import slimeknights.tconstruct.library.tools.helper.ToolBuildHandler;
+import slimeknights.tconstruct.library.tools.item.IModifiableDisplay;
+import slimeknights.tconstruct.library.tools.layout.LayoutSlot;
+import slimeknights.tconstruct.library.tools.part.IToolPart;
+import slimeknights.tconstruct.tools.TinkerTools;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ToolBuildingCategory implements IRecipeCategory<ToolBuildingRecipe> {
+  private static final ResourceLocation BACKGROUND_LOC = TConstruct.getResource("textures/gui/jei/tinker_station.png");
+  private static final Component TITLE = TConstruct.makeTranslation("jei", "tinkering.tool_building");
+  @Getter
+  private final IDrawable icon;
+  @Getter
+  private final IDrawable background;
+  private final IDrawable anvil, slot;
+  private static final int WIDTH = 134;
+  private static final int HEIGHT = 66;
+  private static final int X_OFFSET = -6;
+  private static final int Y_OFFSET = -15;
+  private static final int SLOT_SIZE = 18;
+  private static final int ITEM_SIZE = 16;
+
+  public ToolBuildingCategory(IGuiHelper guiHelper) {
+    this.icon = guiHelper.createDrawableItemStack(TinkerTools.pickaxe.get().getRenderTool());
+    this.background = guiHelper.createDrawable(BACKGROUND_LOC, 122, 77, WIDTH, HEIGHT);
+    this.slot = guiHelper.createDrawable(BACKGROUND_LOC, 24, 14, SLOT_SIZE, SLOT_SIZE);
+    this.anvil = guiHelper.createDrawable(BACKGROUND_LOC, 128, 61, ITEM_SIZE, ITEM_SIZE);
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayoutBuilder builder, ToolBuildingRecipe recipe, IFocusGroup focuses) {
+    List<IToolPart> parts = recipe.getToolParts();
+    List<List<ItemStack>> allParts = recipe.getAllToolParts();
+    List<List<ItemStack>> extras = recipe.getExtraRequirements().stream().map(ingredient -> Arrays.asList(ingredient.getItems())).toList();
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    if (parts.size() + extras.size() > layoutSlots.size()) {
+      TConstruct.LOG.error(String.format("Tool part count is greater than layout slot count for %s!", recipe.getId()));
+      int additionalSlots = 0;
+      layoutSlots = new ArrayList<>(layoutSlots);
+      while (parts.size() + extras.size() > layoutSlots.size()) {
```

Comment:
This could be a for loop which would be more readible. Same below.

```java
int missingSlots = parts.size() + extras.size() - layoutSlots.size()
for (int additionalSlot = 0; additionalSlot < missingSlots; additionalSlot++)
```